[DATA] Identifiers

runspired opened the pull request:
[Rendered](https://github.com/runspired/rfcs/blob/ember-data-identifiers/text/0403-ember-data-identifiers.md)

igorT commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,194 @@
+    * New Re-Arg: `attributesDefinitionFor(identifier)`
+  * `relationshipsDefinitionFor(modelName)`
+    * New Re-Arg: `relationshipsDefinitionFor(identifier)`
+  * `setRecordId(modelName, id, clientId)`
```
Should we make this just `notifyPropertyChange` `id`, no reason to make set id special?

runspired replied:
It is special as we use this to notify that the ID has been received from a save when no ID was present before and that we must now update the cache key / index.

The data flow is *whacked* here though, as `store._setRecordId` which this delegates to is going to call `internalModel.setId` which needs to be fixed to call `recordData.<updateId>` fpr the case where someone calls `store.createRecord('foo', {}).set('id', '<someid>');`

igorT commented on text/0000-ember-data-identifiers.md:
````diff
@@ -0,0 +1,194 @@
+
+```typescript
+/*
+  Returns the RecordIdentifier for the given ResourceIdentifier.
````
Do we need a split between a `recordIdentifier` and a `resourceIdentifier` if we are not guaranteeing stability? If you don't have to pass the given `recordIdentifier`, this just seems like wasted roundtrip

runspired replied:
I think you resolved the difference for yourself below :) one is non-normalized (things we accept from the API), the other is the stricter normalized version (e.g. MUST have `lid` etc.)

dgeb replied:
I would be more comfortable with `ResourceIdentifier` being exported from a JSONAPI-specific package, and not at all part of ember-data core.

runspired replied:
@dgeb I considered that but no available lib matches what we actually expect.

Specifically:

We allow null id, and lid is not in spec yet. We also will have to handle migrating off of clientId to favor lid.

I can see a future in which this doesn’t need to be provided by ember-data though.

hjdivad replied:
@runspired @dgeb can we not make that json-api specific package as part of this change?  presumably with the intent to eventually (or perhaps immediately) move it into the `json-api` org?

igorT commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,194 @@
+
+  - validate the `id` `type` and `lid` combo against known identifiers
+  - return an object that is guaranteed to have the `lid` for the given `type` and `id`
+  - return an object that also contains the stored `meta`
```
This is interesting, haven't thought about it much. Whats the thinking behind attaching the `meta` to it?

runspired replied:
Primarily because the `json-api` spec allows it.

igorT commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,194 @@
+
+**Store Methods**
+* `recordDataFor(modelName, id, clientId)`
+  * New Re-Arg: `recordDataFor(identifier)`
```
Typing the identifier here would help differentiate between the `resourceIdentifier` and `recordIdentifier`

runspired replied:
I will once we have resolved which it should be, this question is in `unresolved questions` atm.

jessica-jordan commented:
@runspired Looking at the WIP tag, is it alright for us - @emberjs/the-ember-times-editors - to feature this already in next week's newsletter?

runspired commented:
@jessica-jordan still too many rough edges, but I think we'll be in good shape by late Wednesday/Thursday for featuring it next week.

hjdivad commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,321 @@
+
+export interface RecordIdentifier {
+    lid: string;
+    id: string | null;
```
Doesn't `id?` make more sense here?

You may not have an id at the time you create a `RecordIdentifier`, in particular when you create the record locally, but that doesn't make the id `null` it makes it unknown.

runspired replied:
an unknown `id` is `null`.

hjdivad commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,321 @@
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
```
similarly i think `meta?` is more appropriate.

runspired replied:
`?` would imply that the key is sometimes `undefined`. To maintain shape, we always have the key present.

hjdivad commented on text/0000-ember-data-identifiers.md:
````diff
@@ -0,0 +1,321 @@
+    type: string;
+    meta: Meta | null;
+}
+```
````
`ResourceIdentifier` and `RecordIdentifier` have extremely similar names that do not make clear their semantic differences (ie that a record identifier is referentially stable and a resource identifier may not be).

What do you think about renaming `RecordIdentifier` to `StableIdentifier`.

runspired replied:
Stability is not the difference between the two.

A `ResourceIdentifier` is the `json-api` interface, which may or may not have `lid` and `meta` but MUST have `type` (and must have `id` if not for a serialized version of a new client created resource with an `lid`)
A `RecordIdentifier` is our own interface for internal use, which additionally MUST have `lid` and which always has `meta` (although it may be `null`).

runspired replied:
cc @dgeb for thoughts, as we are paralleling a similar conceptual parity from orbit here.

runspired replied:
> @runspired @dgeb can we not make that json-api specific package as part of this change? presumably with the intent to eventually (or perhaps immediately) move it into the json-api org?

Not sure why I can't reply to this comment on the comment, but the TL;DR is that it is unlikely we ever perfectly match the signature of jsonapi here for `RecordIdentifier`, but we may be able to match the official interface for `ResourceIdentifier` once the spec includes the `lid` addition.

hjdivad replied:
For the separate package I was referring specifically to `ResourceIdentifier`.  I agree `RecordIdentifier` should be internal.

---

> Stability is not the difference between the two.

> Additionally, our identifiers used internally or given from ember-data to application code will be referentially stable, allowing them to be used as keys in either WeakMap or Map based caches.

Are `RecordIdentifier`s referentially stable?

runspired commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,321 @@
+
+export interface ResourceIdentifier {
+  lid?: string;
+  id: string | null;
```
optional (`?`) has a very different meaning from `string | null`. Here, we are guaranteeing that the key is present, but `null` if not a `string`.

runspired commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,321 @@
+
+In order to rectify this, we would need to add either `identifier` or `modelName, id, clientId` to the front of every method signature provided by `RecordData` as well as many methods provided by `StoreWrapper`. Given our interest in reducing the complexity of method signatures and unifying concepts of `identity`, we feel the path of least churn is to use this opportunity to correct the `RecordDa ...
+
+### Operations
```
It is the path by which we correct the mistakes of the original implementation

hjdivad replied:
That may be, but we can switch to record identifiers without operations, and operations without record identifiers.  It's not a big deal, but tackling both at once makes the rfc less focused IMHO  (and subsequently less clear).

runspired commented on text/0000-ember-data-identifiers.md:
````diff
@@ -0,0 +1,250 @@
+ will be discussed further below.
+
+```typescript
+type GenerationMethod = (data: Object, bucket?: string) => Identifier;
````
fixed

hjdivad commented on text/0000-ember-data-identifiers.md:
````diff
@@ -0,0 +1,396 @@
+
+  return lid;
+});
+ ```
````
It would help to augment this example with store API to make it clear what the flow is.

eg if I'm doing

```js
setIdentifierGenerationMethod(exampleInRFC);

store.findRecord('user', 'jackson5').
  then(doSomething); // this makes a request

// later

store.findRecord('user', 'abc123').
  then(doSomethingWithUser); // this does not make a request
```

At what times does the store call the generation method and with what arguments?

runspired replied:
Users should do any identifier generation configuration within an instance initializer prior to making use of the store.

For `RecordIdentifiers` the method is called once a record (remote or local) becomes known about. In the `findRecord` case this will be greedy (prior to network request being sent), so as long as the first request for `jackson5` in the example had populated a secondary lookup table for `abc123` to point at the same identifier prior to the second request being fired no request would be made.

In cases such as `findRecord` or async relationships where `type+id` is known but full resources are not available, `updateRecordIdentifier` will be called once the resource is loaded to ensure that we populate these secondary lookup tables.

I'll add clarifying text around this.

hjdivad replied:
> In the findRecord case this will be greedy (prior to network request being sent), so as long as the first request for jackson5 in the example had populated a secondary lookup table for abc123 to point at the same identifier prior to the second request being fired no request would be made.

Indeed.  What causes that secondary cache population to happen?  Is your intent in the example above for that to happen out of band from a userland adapter?  If so, adding that detail will go a long way to making it clear how this example solves the issue of secondary caches.

runspired replied:
Added more details around secondary caches and expanded examples.

hjdivad commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,396 @@
+
+In an ideal world, the `lid` of each `Identifier` would be a `v4` [uuid](https://en.wikipedia.org/wiki/Universally_unique_identifier),
+ making it unique in all contexts. However, due to requirements around design flexibility and performance we are only requiring that 
+ `Identifiers` be unique _within their bucket_ for the data they are intended to reference.
```
What about this requirement + multiple stores?  It's something that's been thought about and maybe deserves a few words of clarity in the RFC.

runspired replied:
Great suggestion, will add 👍

runspired replied:
Added

hjdivad commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,396 @@
+  [IS_IDENTIFIER]: true;
+  [DEBUG_IDENTIFIER_BUCKET]: string;
+  lid: string;
+}
```
What came of the discussion around using only `lid: string` and storing additional information in a `WeakMap`?

runspired replied:
As it was discussed further with other folks (including @stefanpenner) the realization was had that we do not need referential stability for WeakMap purposes (both because `lid` is stable and because ember-data has need of controlling most of the object lifecycle).

However, continuing to use an object wrapper, especially a stable object wrapper, comes with a multitude of benefits, including:

- **debugging:** enhanced debugging by associating additional information with the identifier in dev builds
- **debugging:** enhanced debugging by users being able to see `type` and `id` at a glance when
- **bug prevention:** enforcement of the use of the identifier generation process (to ensure lookup tables are properly populated)
- **debugging:** ability to tell at a glance that an identifier was properly processed
- **ergonomics:** closer alignment to jsonapi that makes it true that all `RecordIdentifiers` are also `ResourceIdentifierObjects`

runspired replied:
I added this reasoning into a discussion in `Alternatives`

runspired commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,396 @@
+ `Identifiers` be unique _within their bucket_ for the data they are intended to reference.
+
+Each underlying primitive will have it's own bucket, as new primitives are formalized, new buckets will emerge. Initially we expect
+ only a `resource` bucket which aligns with today's `IdentityMap` cache for `Record`s. Examples of future buckets may include a
```
`resource` is the `string` value that the bucket symbol would be set to in this case.

runspired replied:
(and it should have been the string `record`, fixed)

runspired commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,396 @@
+When this occurs, the existing `lid` needs the chance to be forgotten and a new `lid` generated. This method eliminates
+  the identifier from our internal cache only. Caches associated with custom identifiier generation methods must be cleared
+  by the implementors of those custom methods. Any data associated with the original `lid` should be purged from caches prior
+  to calling this method. We leave it to follow up RFCs to provide infrastructure for safely and correctly eliminating records from caches.
```
We still must handle connecting data coming from the API which may or may not have an `lid` to the appropriate identifier, if the API gives us the same `type+id` combo that we have seen previously, but for a different piece of data we would introduce mistakes anywhere we re-used the same `lid`. 

Perhaps in some situations this is allowable or even preferable, and users implementing custom generation methods can determine this for themselves: however we need to give users a chance to purge their existing caches for the stale `lid`.

hjdivad replied:
fair enough

runspired commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,396 @@
+
+  updateRecordIdentifier(identifier: Identifier, data: Resource): void;
+
+  forgetRecordIdentifier(identifier: Identifier): void
```
`forget` differs significantly from `unload` in semantics and a new verb was chosen here in part to convey intent and partly to avoid the baggage that comes with `unload`.

`forget` is intended to fully purge a given identifier from the identifier cache. It differs in semantics from `unload` primarily in how we consider the existence (or lack thereof) of remote data.

What this means is that we treat the identifier as if we no longer know anything about it locally or remotely. E.g. it may or may not exist remotely still, but we remove all reference to it.

This differs from the semantics of `unload` in which if a record has not first had a deletion persisted we move to a state in which we assume no local data but that remote data does still exist.

hjdivad replied:
ok.  I'm not a fan of the name.  I would recommend `removeRecordIdentifier` mainly b/c `remove` is more standard terminology for caches than `forget`.

I don't plan on being a 1-man bikeshedder though if others feel strongly about `forget`.

runspired commented on text/0000-ember-data-identifiers.md:
````diff
@@ -0,0 +1,396 @@
+```ts
+Store {
+  updateRecordIdentifier(identifier: Identifier, data: Resource): void;
+}
````
It provides the opportunity to update the primary and secondary lookup tables for the identifier.

Additionally, in the case of a `RecordIdentifier` that was created locally, it provides the ability to do a "one time only" upgrade of the identifier to assign an `id`.

hjdivad replied:
> Additionally, in the case of a RecordIdentifier that was created locally, it provides the ability to do a "one time only" upgrade of the identifier to assign an id.

Are the semantics then that it mutates the `identifier` parameter?

hjdivad replied:
Ok so per OOB discussion we mutate the argument here.  That makes sense; I'd recommend making that explicit here.

runspired replied:
Yes. This is now made clear in the RFC

runspired commented on text/0000-ember-data-identifiers.md:
````diff
@@ -0,0 +1,396 @@
+setIdentifierGenerationMethod((resource: Resource) => {
+  return resource.lid || `my-key-${count++}`;
+});
+```
````
Great suggestion 👍

dgeb commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,651 @@
+  lid?: string;
+  id: string;
+  type: string;
+}
```
Is there a reason to include `ResourceIdentifierObject` here? This should be an adapter/serailizer concern that seems to muddy the clarity of this proposal, which deals with records and their identity.

dgeb commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,658 @@
+}
+
+export interface RecordIdentifier extends Identifier {
+  id: string | null;
```
I believe this is more properly `id?: string` since `id` may be completely missing (which seems preferable to `null`).

runspired replied:
`null` is correct as that is what new records will have; however, a missing id when `lid` is available is allowable, so updated to allow not having `id` in this case.

dgeb commented on text/0000-ember-data-identifiers.md:
```diff
@@ -0,0 +1,658 @@
+  [DEBUG_IDENTIFIER_BUCKET]: string;
+}
+
+export interface RecordIdentifier extends StableIdentifier {
```
Shouldn't this be `StableRecordIdentifier` ?

runspired replied:
Thanks for the catch! Over eager refactor 😂

courajs commented:
I know I'm pretty late to the party here, but are we sure we don't want to take the opportunity to generalize away from string-only ids? Sometimes it is useful to use compound identifiers. I'm actually using them in an (experimental, without ED) app right now - `{site_id: '<a uuid>', index: <an integer>}`.

Although these can be serialized to/from strings, it places restrictions on the individual sub-keys (no colons, or must remain a fixed length, etc).

Maybe the cost is too great (need to define an equality protocol, makes maps/caches/lookups harder & more expensive, etc), but I'd love to hear any thoughts on the subject.

runspired commented:
@courajs 

> I know I'm pretty late to the party here, but are we sure we don't want to take the opportunity to generalize away from string-only ids? Sometimes it is useful to use compound identifiers. I'm actually using them in an (experimental, without ED) app right now - {site_id: '<a uuid>', index: <an integer>}.

Compound identifiers still need to be serializable and (more importantly) would not provide a path for resolving a large number of the edge cases this RFC tackles (mainly around polymorphism, alternate indexes, new record creation, and race conditions during save).

Moreover (and not to nitpick, but I feel this is important) in the example compound key you gave (`{site_id: '<a uuid>', index: <an integer>}`), only one of the two properties is actually required. If `site_id` is truly a `uuid`, then it is unique across types and not merely within the space that `index` represents. Obviously, the more common case here is that `site_id` is not a `uuid`.

> Although these can be serialized to/from strings, it places restrictions on the individual sub-keys (no colons, or must remain a fixed length, etc).

I feel just the opposite is true here.

First: we don't attach an arbitrary length to the `lid` requirement, nor does it propose any meaning around delimiters within the string (such as colons). All we ask is that given some data, you tell us how to identify. The specifics around what guarantees you need to provide on key-uniqueness are up to you.

Second: While you _could_ attempt to parse your compound properties out of the `lid`, you very much _should not_, and we intend to obfuscate the `lid` we generate a bit to help folks avoid trying to parse `type` and `id` information out of our default strategy (vs just using `type` and `id` which are also available for instance).

> Maybe the cost is too great (need to define an equality protocol, makes maps/caches/lookups harder & more expensive, etc), but I'd love to hear any thoughts on the subject.

It would most definitely make a lot of costs greater, lookups harder, equality checks difficult. More importantly it would make serializing the state of things unnecessarily difficult.

Consider the case where the compound key used to generate the `lid` is not 1 or 2 keys but N keys. Ultimately for us to ask you if your object is arbitrarily equal to another object that may not have all N keys would be impossible. However, asking you if an object with a serialized form of identity is equal to an object with all N keys but where the serialized form is not present is very achievable.

sandstrom commented:
Very excited about this! As someone who has followed the `lid` discussion in JSON-API for a while, I think this would be a great improvement to Ember Data! 🏅

runspired commented:
At our meeting yesterday we voted to merge this RFC. Having squashed the commits and renamed the file with the RFC # and updated the appropriate links, I am now merging :)
